kuma/wiki/forms.py

```diff
@@ -233,47 +250,55 @@ def __init__(self, *args, **kwargs):
         if self.section_id:
             self.fields['toc_depth'].required = False
 
-    def _clean_collidable(self, name):
-        value = self.cleaned_data[name]
-
+    def clean_slug(self):
+        # Since this form can change the URL of the page on which the editing
+        # happens, changes to the slug are ignored for an iframe submissions
         if self.is_iframe_target:
-            # Since these collidables can change the URL of the page, changes
-            # to them are ignored for an iframe submission
-            return getattr(self.instance.document, name)
+            return self.instance.document.slug
+
+        # Get the cleaned slug
+        slug = self.cleaned_data['slug']
+
+        # first check if the given slug doesn't contain slashes and other
+        # characters not allowed in a revision slug component (without parent)
+        if slug and INVALID_REV_SLUG_CHARS_RE.search(slug):
+            raise forms.ValidationError(SLUG_INVALID)
+
+        # edits can come in without a slug, so default to the current doc slug
+        if not slug:
+            try:
+                slug = self.instance.slug = self.instance.document.slug
+            except ObjectDoesNotExist:
+                pass
+
+        # then if there is a parent document we prefix the slug with its slug
+        if self.parent_slug:
+            slug = u'/'.join([self.parent_slug, slug])
 
-        error_message = {'slug': SLUG_COLLIDES}.get(name, OTHER_COLLIDES)
         try:
-            existing_doc = Document.objects.get(
-                locale=self.instance.document.locale,
-                **{name: value})
+            doc = Document.objects.get(locale=self.instance.document.locale,
```

Comment:
`self.instance.document` can raise a `RelatedObjectDoesNotExist` which is a subclass of `AttributeError` and `ObjectDoesNotExist`. I think that should only ever happen if we have faulty data in the database, when a revision didn't get a document assigned for some reason